### PR TITLE
BUG: Groupby(sort=False) with datetime-like Categorical raises ValueError

### LINKED ISSUE #10505: BUG: Groupby(sort=False) with datetime-like Categorical raises ValueError
Related to #10501, but not the same. ``groupby`` can accept ``Categorical`` and ``sort`` keyword.

```
df = pd.DataFrame({'A': [1, 2, 3 ,4], 'B': [5, 6, 7, 8]})

# OK
df.groupby(pd.Categorical(['A', 'B', 'A', 'B'])).groups
# {'A': [0, 2], 'B': [1, 3]}

# OK
df.groupby(pd.Categorical(['A', 'B', 'A', 'B']), sort=False).groups
# {'A': [0, 2], 'B': [1, 3]}
```

If ``Categorical`` has datetime-like categories, groupby fails if ``sort=False`` is specified.

```
# OK
df.groupby(pd.Categorical(pd.DatetimeIndex(['2011', '2012', '2011', '2012']))).groups
# {numpy.datetime64('2011-01-01T09:00:00.000000000+0900'): [0, 2], 
#  numpy.datetime64('2012-01-01T09:00:00.000000000+0900'): [1, 3]}

# NG
df.groupby(pd.Categorical(pd.DatetimeIndex(['2011', '2012', '2011', '2012'])), sort=False).groups
# ValueError: items in new_categories are not the same as in old categories
```

### PR DESCRIPTION
**Pull request by sinhrks**
Closes #10505.

### CODE REVIEW

File: pandas/core/groupby.py [hunk 1]
```diff
@@ -1950,7 +1950,12 @@ def __init__(self, index, grouper=None, obj=None, name=None, level=None,
 
                 # fix bug #GH8868 sort=False being ignored in categorical groupby
                 else:
-                    self.grouper = self.grouper.reorder_categories(self.grouper.unique())
+                    from pandas.core.nanops import unique1d
+                    # take unique codes preserving its original order
+                    unique_codes = unique1d(self.grouper.codes)
+                    # do not coerce to np.array to make datetime-like work
+                    cat = self.grouper.categories.take(unique_codes)
+                    self.grouper = self.grouper.reorder_categories(cat)
```
Comment (jreback) [inline pandas/core/groupby.py]: lets make a method in categorical to do like this (and just call it here)

Comment (jreback) [inline pandas/core/groupby.py]: better yet I think unique for a cat should just do it this way

Comment (sinhrks) [inline pandas/core/groupby.py]: How about adding  ``return_type`` kw to allow ``Categorical.unique`` to return ``Categorical`` like above?

Comment (jreback) [inline pandas/core/groupby.py]: I think unique should always return a Categorical

(like above) - wonder why we don't have that

Comment (sinhrks) [inline pandas/core/groupby.py]: Understood. Will confirm whether the API change has any side effect.

**Comment by sinhrks** [comment]
Fixing ``Categorical.unique`` to return unique ``Categorical``. Would like to confirm the expected behavior based on current impl.

- unordered category: values and categories are sorted by appearance order, excluding unnecessary ones.
- ordered category: values are sorted by appearance order, categories keeps existing order excluding unnecessary ones.

```
import pandas as pd
import numpy as np

# unordered
cat = pd.Categorical(["b","b","a","a"], categories=["a","b","c"])
cat.unique()
# [b, a]
# Categories (2, object): [b, a]

cat = pd.Categorical(["b", "b", np.nan, "a"], categories=["a","b","c"]) 
cat.unique()
# [b, NaN, a]
# Categories (2, object): [b, a]

# ordered
cat = pd.Categorical(["b","b","a","a"], categories=["a","b","c"], ordered=True)
cat.unique()
# [b, a]
# Categories (2, object): [a < b]

cat = pd.Categorical(["b", "b", np.nan, "a"], categories=["a","b","c"], ordered=True) 
cat.unique()
# [b, NaN, a]
# Categories (2, object): [a < b]
```

**Comment by jreback** [comment]
the uniques should not have ``nan`` in them. Further I think the ``ordered=True`` should order according to the categories (e.g. the present ones), this might be a small API change.

**Comment by sinhrks** [comment]
Ah, correct. Fixed and updated the example above.

I understand ``groupby`` result must follows the same rule as above. It means ``groupby(Categorical(ordered=True), sorted=False)`` will have the index sorted by the order of categories, was previously ordered by appearance of values (unsorted).

**Comment by sinhrks** [comment]
> uniques should not have nan

I have misunderstand this with ``value_counts``. Currently, ``unique`` returns array including ``NaN`` both normal and categorical.

```
pd.Series([1., 2., 1., np.nan]).unique()
# array([  1.,   2.,  nan])
pd.Series([1., 2., 1., np.nan], dtype='category').unique()
# array([  1.,   2.,  nan])
```

So, the fixed categorical result should preserve ``NaN`` in codes and not in categories as below? 

```
>>> pd.Series([1., 2., 1., np.nan], dtype='category').unique()
# [1, 2, NaN]
# Categories (2, float64): [1, 2]
```

**Comment by jreback** [comment]
lgtm. merge when ready.

**Comment by sinhrks** [comment]
Thanks, let me merge this once rebase and gets green.

**Pull request merged**